### PR TITLE
[SPARK-44264][PYTHON]E2E Testing for Deepspeed

### PR DESCRIPTION
**Pull request by mathewjacob1002**
### What Changes Were Proposed 
This PR adds some end to end tests for the DeepspeedTorchDistributor. Due to the lack of support currently available to deepspeed, we use proxy functions that are very simple to test that the command works. For the future, once Deepspeed supports more CPU, these end to end tests should instead migrate to more actual Deepspeed workloads, such as those described in the PR comments.

### Why Do We Need These Changes
Previously, we only had unit tests for helper functions. These test actual workloads that a user may use DeepspeedTorchDistributor from end to end.

### Any User Facing Changes
No, these are end to end tests.

### How Was This Tested
Running the tests and seeing if they pass.

### CODE REVIEW

**Comment by mathewjacob1002** [comment]
Note to self: The following error occurs from the distributed e2e test: Resource script <temp_file_name> to discover gpu doesn't exist!. After commenting out the code, it is involving the Local and Distributed E2E environments I think. Potentially a problem with the setup/teardown of the classes?

File: python/pyspark/ml/deepspeed/deepspeed_distributor.py [hunk 1]
```diff
@@ -61,6 +61,30 @@ def __init__(
...
+        >>> output = distributor.run(train, 0.01)
+
+        Run Deepspeed training function on multiple nodes
+
+        >>> distributor = DeepspeedTorchDistributor(num_gpus=4,
+                                                    nnodes=3,
+                                                    use_gpu=True,
+                                                    local_mode=False,
+                                                    deepspeed_config="path/to/config.json")
+        >>> output = distributor.run(train, 0.01)
```
Comment (maddiedawson) [inline python/pyspark/ml/deepspeed/deepspeed_distributor.py:87]: Remove this change since it's in the other PR

File: python/pyspark/ml/torch/distributor.py [hunk 2]
```diff
@@ -609,7 +609,7 @@ def _run_local_training(
                     del os.environ[CUDA_VISIBLE_DEVICES]
 
         return output
-
+
```
Comment (maddiedawson) [inline python/pyspark/ml/torch/distributor.py:612]: Undo this change

File: python/pyspark/ml/deepspeed/deepspeed_distributor.py [hunk 3]
```diff
@@ -127,6 +151,10 @@ def _run_training_on_pytorch_file(
         training_command = DeepspeedTorchDistributor._create_torchrun_command(
             input_params, train_path, *args
         )
+        # Spark CI doesn't have GPUs (primary use case for deepspeed); 
+        # Just make sure everything else would work
+        if TorchDistributor._E2E_MOCK:
+            return training_command
```
Comment (maddiedawson) [inline python/pyspark/ml/deepspeed/deepspeed_distributor.py:157]: Why does this need to return anything?

File: python/pyspark/ml/deepspeed/tests/test_deepspeed_distributor.py [hunk 4]
```diff
@@ -18,9 +18,14 @@
 import sys
 from typing import Any, Tuple, Dict
 import unittest
+import shutil
```
Comment (maddiedawson) [inline python/pyspark/ml/deepspeed/tests/test_deepspeed_distributor.py:21]: Alphabetize these

File: python/pyspark/ml/deepspeed/tests/test_deepspeed_distributor.py [hunk 5]
```diff
@@ -163,6 +168,93 @@ def test_create_torchrun_command_distributed(self) -> None:
             )
             self.assertEqual(distributed_cmd_args_expected, distributed_command_with_args)
 
+def _create_basic_function():
+   def pythagoras(leg1: float, leg2: float):
+       import deepspeed
+       return (leg1 * leg1 + leg2 * leg2)**0.5
+   return pythagoras 
+
+from pyspark.ml.torch.tests.test_distributor import get_distributed_mode_conf
```
Comment (maddiedawson) [inline python/pyspark/ml/deepspeed/tests/test_deepspeed_distributor.py:177]: Put this with the other imports

File: python/pyspark/ml/deepspeed/tests/test_deepspeed_distributor.py [hunk 6]
```diff
@@ -163,6 +168,93 @@ def test_create_torchrun_command_distributed(self) -> None:
...
+
+class DeepspeedTorchDistributorDistributedEndToEnd(unittest.TestCase):
+
+    @classmethod
+    def setUpClass(cls):
+        (cls.gpu_discovery_script_file_name, cls.mnist_dir_path) = set_up_test_dirs()
+        print(f"Distributed e2e fname: {cls.gpu_discovery_script_file_name}")
```
Comment (maddiedawson) [inline python/pyspark/ml/deepspeed/tests/test_deepspeed_distributor.py:184]: Let's remove the print statements

File: python/pyspark/ml/deepspeed/tests/test_deepspeed_distributor.py [hunk 7]
```diff
@@ -163,6 +168,93 @@ def test_create_torchrun_command_distributed(self) -> None:
...
+        conf = SparkConf(loadDefaults=False)
+        print("BEFORE setting anything: ", conf.getAll(), " |||||||||||\n")
+        for k, v in get_distributed_mode_conf().items():
+            if k is "spark.driver.resource.gpu.discoveryScript":
+                raise RuntimeError("Found trying to set the driver discovery script")
```
Comment (maddiedawson) [inline python/pyspark/ml/deepspeed/tests/test_deepspeed_distributor.py:189]: Do we need this?

File: python/pyspark/ml/deepspeed/tests/test_deepspeed_distributor.py [hunk 8]
```diff
@@ -163,6 +168,93 @@ def test_create_torchrun_command_distributed(self) -> None:
...
+            conf = conf.set(k, v)
+        conf = conf.set(
+            "spark.worker.resource.gpu.discoveryScript", cls.gpu_discovery_script_file_name
+        )
+        print("Distributed babby:", conf.getAll())
+        sc = SparkContext("local-cluster[2,2,512]",cls.__name__,conf=conf)
```
Comment (maddiedawson) [inline python/pyspark/ml/deepspeed/tests/test_deepspeed_distributor.py:195]: How did you choose 2,2,512?

File: python/pyspark/ml/torch/distributor.py [hunk 9]
```diff
@@ -918,6 +918,13 @@ def _create_save_dir(root_dir: Optional[str] = None) -> str:
     @staticmethod
     def _cleanup_files(save_dir: str) -> None:
         shutil.rmtree(save_dir, ignore_errors=True)
+    
+    @staticmethod
+    @contextmanager
+    def _setup_e2e_mocking_env():
+        TorchDistributor._E2E_MOCK = True
+        yield
+        TorchDistributor._E2E_MOCK = False
```
Comment (maddiedawson) [inline python/pyspark/ml/torch/distributor.py:927]: How is this used?

Comment (mathewjacob1002) [reply python/pyspark/ml/torch/distributor.py:927]: this is old code - will remove now

File: python/pyspark/ml/deepspeed/deepspeed_distributor.py (hunk 3 above)
Comment (mathewjacob1002) [reply python/pyspark/ml/deepspeed/deepspeed_distributor.py:157]: old code before figuring out stuff with the SparkConf - will clean up now

File: python/pyspark/ml/deepspeed/tests/test_deepspeed_distributor.py (hunk 8 above)
Comment (mathewjacob1002) [reply python/pyspark/ml/deepspeed/tests/test_deepspeed_distributor.py:195]: This is the same as the TorchDistributor tests (https://github.com/apache/spark/blob/master/python/pyspark/ml/torch/tests/test_distributor.py). It's just to set up the env so that Spark doesn't complain.

File: python/pyspark/ml/deepspeed/tests/test_deepspeed_distributor.py [hunk 10]
```diff
@@ -17,10 +17,16 @@
 import os
 import sys
 from typing import Any, Tuple, Dict
+
```
Comment (maddiedawson) [inline python/pyspark/ml/deepspeed/tests/test_deepspeed_distributor.py:20]: Remove extra newline and alphabetize this

File: python/test_support/test_deepspeed_training_file.py [hunk 11]
```diff
@@ -0,0 +1,10 @@
+import sys
+
+def pythagorean_thm(x : int, y: int):
+    import deepspeed
+    return (x*x + y*y)**0.5
+
+def main():
```
Comment (maddiedawson) [inline python/test_support/test_deepspeed_training_file.py:7]: Don't need the main() function, you can just call print(pythagorean_thm(...))

File: python/pyspark/ml/deepspeed/tests/test_deepspeed_distributor.py [hunk 12]
```diff
@@ -163,6 +169,84 @@ def test_create_torchrun_command_distributed(self) -> None:
...
+       return (leg1 * leg1 + leg2 * leg2)**0.5
+   return pythagoras 
+
+
+class DeepspeedTorchDistributorDistributedEndToEnd(unittest.TestCase):
+
+    @classmethod
+    def setUpClass(cls):
+        (cls.gpu_discovery_script_file_name, cls.mnist_dir_path) = set_up_test_dirs()
+        conf = SparkConf(loadDefaults=False)
```
Comment (maddiedawson) [inline python/pyspark/ml/deepspeed/tests/test_deepspeed_distributor.py:184]: Add a comment explaining why we use loadDefaults=False

File: python/pyspark/ml/deepspeed/tests/test_deepspeed_distributor.py [hunk 13]
```diff
@@ -163,6 +169,84 @@ def test_create_torchrun_command_distributed(self) -> None:
...
+       train_fn = _create_basic_function()
+       # Arguments for the pythagoras function train_fn
+       x = 3
+       y = 4
+       dist = DeepspeedTorchDistributor(num_gpus=2, use_gpu=False, local_mode=False)
+       output = dist.run(train_fn, x, y)
+       self.assertEqual(output, 5)
+
+    def test_pytorch_file_e2e(self):
+        path_to_train_file = "python/test_support/test_deepspeed_training_file.py"
```
Comment (maddiedawson) [inline python/pyspark/ml/deepspeed/tests/test_deepspeed_distributor.py:203]: This will be an issue if someone runs the test from a different directory. Instead of committing python/test_support/test_deepspeed_training_file.py, let's just write out its contents to the /tmp file in the test setup or test case

**Review by maddiedawson — COMMENTED** [review]
Add a TODO in here to use a real training function and explain why we can't have it right now. Also in that comment, you can provide an example of a simple training function to test with

File: python/pyspark/ml/deepspeed/tests/test_deepspeed_distributor.py (hunk 12 above)
Comment (maddiedawson) [reply python/pyspark/ml/deepspeed/tests/test_deepspeed_distributor.py:195]: "This is set to False" -> "loadDefaults is set to False"

File: python/pyspark/ml/deepspeed/tests/test_deepspeed_distributor.py [hunk 14]
```diff
@@ -163,6 +169,105 @@ def test_create_torchrun_command_distributed(self) -> None:
...
+       # Arguments for the pythagoras function train_fn
+       x = 3
+       y = 4
+       dist = DeepspeedTorchDistributor(num_gpus=2, use_gpu=False, local_mode=True)
+       output = dist.run(train_fn, x, y)
+       self.assertEqual(output, 5)
+
+
+    def test_pytorch_file_e2e(self):
+        path_to_train_file = "python/test_support/test_deepspeed_training_file.py"
```
Comment (maddiedawson) [inline python/pyspark/ml/deepspeed/tests/test_deepspeed_distributor.py:262]: This needs to be updated

File: python/pyspark/ml/deepspeed/tests/test_deepspeed_distributor.py [hunk 15]
```diff
@@ -163,6 +169,105 @@ def test_create_torchrun_command_distributed(self) -> None:
...
+        dist = DeepspeedTorchDistributor(num_gpus=2, use_gpu=False, local_mode=True)
+        dist.run(path_to_train_file, 2, 5)
+
+    @classmethod
+    def tearDownClass(cls):
+        shutil.rmtree(cls.mnist_dir_path)
+        os.unlink(cls.gpu_discovery_script_file_name)
+        cls.spark.stop()
```
Comment (maddiedawson) [inline python/pyspark/ml/deepspeed/tests/test_deepspeed_distributor.py:270]: Both here and above, tearDownClass should appear next to setUpClass

**Review by maddiedawson — COMMENTED** [review]
Please add an example training function in the TODO

File: python/pyspark/ml/deepspeed/tests/test_deepspeed_distributor.py [hunk 16]
```diff
@@ -14,13 +14,21 @@
...
+import textwrap
 from typing import Any, Tuple, Dict
+
+import shutil
 import unittest
 
+from pyspark import SparkConf, SparkContext
 from pyspark.ml.deepspeed.deepspeed_distributor import DeepspeedTorchDistributor
-
+from pyspark.ml.torch.distributor import TorchDistributor
```
Comment (maddiedawson) [inline python/pyspark/ml/deepspeed/tests/test_deepspeed_distributor.py:28]: Remove this? The TorchDistributor itself doesn't seem to be used

File: python/pyspark/ml/deepspeed/tests/test_deepspeed_distributor.py [hunk 17]
```diff
@@ -14,13 +14,21 @@
...
+from pyspark.sql import SparkSession
+from pyspark.testing.utils import SPARK_HOME
```
Comment (maddiedawson) [inline python/pyspark/ml/deepspeed/tests/test_deepspeed_distributor.py:30]: Remove, unused

**Review by maddiedawson — APPROVED** [review]

**Comment by mathewjacob1002** [comment]
cc @jeffra in case you wanted to take a look :)

**Review by lu-wang-dl — APPROVED** [review]

File: python/pyspark/ml/deepspeed/tests/test_deepspeed_distributor.py [hunk 18]
```diff
@@ -164,6 +174,178 @@ def test_create_torchrun_command_distributed(self) -> None:
...
+        print(deepspeed.__version__)
+        return (leg1 * leg1 + leg2 * leg2) ** 0.5
+
+    return pythagoras
+
+
+@contextmanager
+def _create_pytorch_training_test_file():
+    # TODO: swap with better training file like below:
+    # import deepspeed
```
Comment (HyukjinKwon) [inline python/pyspark/ml/deepspeed/tests/test_deepspeed_distributor.py:192]: should better move commented codes. We can add them back when we actually need.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.5.

**Comment by zhengruifeng** [comment]
following tests are actually skipped:
```
Skipped tests in pyspark.ml.deepspeed.tests.test_deepspeed_distributor with python3.9:
      test_pytorch_file_e2e (pyspark.ml.deepspeed.tests.test_deepspeed_distributor.DeepspeedDistributorLocalEndToEndTests) ... skip (0.001s)
      test_simple_function_e2e (pyspark.ml.deepspeed.tests.test_deepspeed_distributor.DeepspeedDistributorLocalEndToEndTests) ... skip (0.001s)
      test_pytorch_file_e2e (pyspark.ml.deepspeed.tests.test_deepspeed_distributor.DeepspeedTorchDistributorDistributedEndToEnd) ... skip (0.000s)
      test_simple_function_e2e (pyspark.ml.deepspeed.tests.test_deepspeed_distributor.DeepspeedTorchDistributorDistributedEndToEnd) ... skip (0.000s)
```

due to the `deepspeed` was not installed.

The OSS CI doesn't use `dev/requirements.txt` to install dependency, you may install it [here](https://github.com/apache/spark/blob/46491403bcc794a889103bc0a13329d06454227f/dev/infra/Dockerfile#L73-L75)

**Comment by HyukjinKwon** [comment]
@mathewjacob1002 and @maddiedawson can you follow up ^ please?
